changelog_gen/extractor.py

```diff
@@ -21,37 +20,66 @@ def __init__(self: typing.Self, supported_sections: list[str], *, dry_run: bool
-            if issue.is_file and not issue.name.startswith("."):
-                issue_ref, section = issue.name.split(".")
-                section = section_mapping.get(section, section)
-
-                breaking = False
-                if section.endswith("!"):
-                    section = section[:-1]
-                    breaking = True
-
-                contents = issue.read_text().strip()
-                if section not in self.supported_sections:
-                    msg = f"Unsupported CHANGELOG section {section}"
-                    raise errors.InvalidSectionError(msg)
+        sections = defaultdict(dict)
+
+        if self.has_release_notes:
+            # Extract changelog details from release note files.
+            for issue in sorted(self.release_notes.iterdir()):
+                if issue.is_file and not issue.name.startswith("."):
+                    issue_ref, section = issue.name.split(".")
+                    section = section_mapping.get(section, section)
+
+                    breaking = False
+                    if section.endswith("!"):
+                        section = section[:-1]
+                        breaking = True
+
+                    contents = issue.read_text().strip()
+                    if section not in self.supported_sections:
+                        msg = f"Unsupported CHANGELOG section {section}"
+                        raise errors.InvalidSectionError(msg)
+
+                    sections[section][issue_ref] = {
+                        "description": contents,
+                        "breaking": breaking,
+                    }
+
+        latest_info = Git.get_latest_tag_info()
+        logs = Git.get_logs(latest_info["current_tag"])
+
+        # Build a conventional commit regex based on configured sections
+        #   ^(build|chore|ci|docs|feat|fix|perf|refactor|revert|style|test){1}(\([\w\-\.]+\))?(!)?: ([\w ])+([\s\S]*)
+        types = "|".join(self.supported_sections.keys())
+        reg = re.compile(f"^({types}){{1}}(\([\w\-\.]+\))?(!)?: ([\w .]+)+([\s\S]*)")  # noqa: W605
+
+        for i, log in enumerate(logs):
+            m = reg.match(log)
+            if m:
+                section = m[1]
+                scope = m[2]
+                breaking = m[3] is not None
+                message = m[4]
+                details = m[5] or ""
+
+                # Handle missing refs in commit message, skip link generation in writer
+                issue_ref = f"__{i}__"
+                breaking = breaking or "BREAKING CHANGE" in details
+                for line in details.split("\n"):
+                    m = re.match(r"Refs: #?([\w-]+)", line)
+                    if m:
+                        issue_ref = m[1]
 
                 sections[section][issue_ref] = {
-                    "description": contents,
+                    "description": message,
                     "breaking": breaking,
+                    "scope": scope,
                 }
```

Comment:
🙃 : move to its own method that updates a given `sections` dict?

Reply:
Will check later PR, I believe this was done. Definitely the separate method

Reply:
https://github.com/EdgyEdgemond/changelog-gen/issues/94